contracts/staking/TokenStaking.sol

```diff
@@ -0,0 +1,1236 @@
+            applicationInfo[application].panicButton == msg.sender,
+            "Caller is not the address of panic button"
+        );
+        _;
+    }
+
+    modifier onlyAuthorizerOf(address operator) {
+        require(
+            operators[operator].authorizer == msg.sender,
+            "Not operator authorizer"
+        );
+        _;
+    }
+
+    /// @param _token Address of T token contract
+    /// @param _keepStakingContract Address of Keep staking contract
+    /// @param _nucypherStakingContract Address of NuCypher staking contract
+    /// @param _keepVendingMachine Address of Keep vending machine
+    /// @param _nucypherVendingMachine Address of NuCypher vending machine
+    constructor(
+        T _token,
+        IKeepTokenStaking _keepStakingContract,
+        INuCypherStakingEscrow _nucypherStakingContract,
+        VendingMachine _keepVendingMachine,
+        VendingMachine _nucypherVendingMachine
+    ) {
+        require(
+            _token.totalSupply() > 0 &&
+                _keepStakingContract.ownerOf(address(0)) == address(0) &&
+                _nucypherStakingContract.getAllTokens(address(0)) == 0,
+            "Wrong input parameters"
+        );
+        token = _token;
+        keepStakingContract = _keepStakingContract;
+        nucypherStakingContract = _nucypherStakingContract;
+
+        keepFloatingPointDivisor = _keepVendingMachine.FLOATING_POINT_DIVISOR();
+        keepRatio = _keepVendingMachine.ratio();
+        nucypherFloatingPointDivisor = _nucypherVendingMachine
+            .FLOATING_POINT_DIVISOR();
+        nucypherRatio = _nucypherVendingMachine.ratio();
+    }
+
+    //
+    //
+    // Delegating a stake
+    //
+    //
+
+    /// @notice Creates a delegation with `msg.sender` owner with the given
+    ///         operator, beneficiary, and authorizer. Transfers the given
+    ///         amount of T to the staking contract.
+    /// @dev The owner of the delegation needs to have the amount approved to
+    ///      transfer to the staking contract.
+    function stake(
+        address _operator,
+        address payable _beneficiary,
+        address _authorizer,
+        uint96 _amount
+    ) external override {
+        require(_operator != address(0), "Operator must be specified");
+        OperatorInfo storage operator = operators[_operator];
+        (, uint256 createdAt, ) = keepStakingContract.getDelegationInfo(
+            _operator
+        );
+        require(
+            createdAt == 0 && operator.owner == address(0),
+            "Operator is already in use"
+        );
+        require(
+            _amount > minTStakeAmount,
+            "Amount to stake must be greater than minimum"
+        );
+        operator.owner = msg.sender;
+        operator.authorizer = _authorizer != address(0)
+            ? _authorizer
+            : msg.sender;
+        operator.beneficiary = _beneficiary != address(0)
+            ? _beneficiary
+            : payable(msg.sender);
```

Comment:
Why not to `require` they need to be provided? They need to be passed as function parameters anyway.